src/integrate/ensemble_pimd.cu

```diff
@@ -69,22 +69,33 @@ void Ensemble_PIMD::initialize(Atom& atom)
   std::vector<double*> force_beads_cpu(number_of_beads);
   std::vector<double*> virial_beads_cpu(number_of_beads);
 
-  atom.position_beads.resize(number_of_beads);
-  atom.velocity_beads.resize(number_of_beads);
-  atom.potential_beads.resize(number_of_beads);
-  atom.force_beads.resize(number_of_beads);
-  atom.virial_beads.resize(number_of_beads);
+  if (atom.number_of_beads == 0) {
+    if (!thermostat_centroid) {
+      PRINT_INPUT_ERROR("Cannot use RPMD or TRPMD before PIMD\n.");
+    }
+    atom.position_beads.resize(number_of_beads);
+    atom.velocity_beads.resize(number_of_beads);
+    atom.potential_beads.resize(number_of_beads);
+    atom.force_beads.resize(number_of_beads);
+    atom.virial_beads.resize(number_of_beads);
+  } else {
+    if (atom.number_of_beads != number_of_beads) {
+      PRINT_INPUT_ERROR("Cannot change the number of beads for PIMD runs\n.");
+    }
+  }
 
   for (int k = 0; k < number_of_beads; ++k) {
-    atom.position_beads[k].resize(number_of_atoms * 3);
-    atom.velocity_beads[k].resize(number_of_atoms * 3);
-    atom.potential_beads[k].resize(number_of_atoms);
-    atom.force_beads[k].resize(number_of_atoms * 3);
-    atom.virial_beads[k].resize(number_of_atoms * 9);
-
-    atom.position_beads[k].copy_from_device(atom.position_per_atom.data());
-    atom.velocity_beads[k].copy_from_device(atom.velocity_per_atom.data());
-    atom.force_beads[k].copy_from_device(atom.force_per_atom.data());
+    if (atom.number_of_beads == 0) {
+      atom.position_beads[k].resize(number_of_atoms * 3);
+      atom.velocity_beads[k].resize(number_of_atoms * 3);
+      atom.potential_beads[k].resize(number_of_atoms);
+      atom.force_beads[k].resize(number_of_atoms * 3);
+      atom.virial_beads[k].resize(number_of_atoms * 9);
+
+      atom.position_beads[k].copy_from_device(atom.position_per_atom.data());
+      atom.velocity_beads[k].copy_from_device(atom.velocity_per_atom.data());
+      atom.force_beads[k].copy_from_device(atom.force_per_atom.data());
+    }
```

Comment:
Why does this need to be wrapped in an if-statement? If `number_of_beads` is 0, would this for loop even execute (since the for-statement is falsy, 0 < 0)?

Reply:
> Why does this need to be wrapped in an if-statement? If `number_of_beads` is 0, would this for loop even execute (since the for-statement is falsy, 0 < 0)?
> 
> Here `atom.number_of_beads` and `number_of_beads` are two independent variables. `number_of_beads` must be nonzero up to this point and `atom.number_of_beads` will be zero for the first PIMD run and it will only be equal to `number_of_beads` at line 107 below.

I "edited" again :-)

Reply:
The point is that these atom.xxx_beads vectors only need to be allocated and initialized once for the first PIMD run. For later PIMD-related runs, they should not be re-initialized.

Reply:
Ah I see, thanks for the clarification!